### PR TITLE
network: obtain address instance from fd, bind-to-zero in ListenSocke…

### PR DESCRIPTION
**Pull request by htuch**
…tImpl.

Refactor 33ece570 to assist with zero binding from integration tests.

### CODE REVIEW

**Comment by htuch** [comment]
@jamessynge

File: source/common/network/address_impl.cc [hunk 1]
```diff
@@ -8,6 +8,37 @@
 namespace Network {
 namespace Address {
 
+InstanceConstSharedPtr addressFromFd(int fd) {
+  sockaddr_storage ss;
+  socklen_t ss_len = sizeof ss;
+  const int rc = ::getsockname(fd, reinterpret_cast<sockaddr*>(&ss), &ss_len);
+  if (rc != 0) {
+    const int err = errno;
```
Comment (mattklein123) [inline source/common/network/address_impl.cc]: nit: local var not needed

File: source/common/network/address_impl.cc [hunk 2]
```diff
@@ -8,6 +8,37 @@
...
+  case AF_INET6: {
+    ASSERT(ss_len == sizeof(sockaddr_in6));
+    const struct sockaddr_in6* sin6 = reinterpret_cast<const struct sockaddr_in6*>(&ss);
+    ASSERT(AF_INET6 == sin6->sin6_family);
+    return InstanceConstSharedPtr(new Address::Ipv6Instance(*sin6));
+    break;
+  }
+  default:
+    throw EnvoyException(fmt::format("Unexpected family in getsockname result: {}", ss.ss_family));
+    return nullptr;
```
Comment (mattklein123) [inline source/common/network/address_impl.cc]: nit: line can't be reached.

File: source/common/network/address_impl.cc [hunk 3]
```diff
@@ -8,6 +8,37 @@
...
+    throw EnvoyException(fmt::format("getsockname failed for '{}': {}", fd, strerror(err)));
+    return nullptr;
+  }
+  switch (ss.ss_family) {
+  case AF_INET: {
+    ASSERT(ss_len == sizeof(sockaddr_in));
+    const struct sockaddr_in* sin = reinterpret_cast<const struct sockaddr_in*>(&ss);
+    ASSERT(AF_INET == sin->sin_family);
+    return InstanceConstSharedPtr(new Address::Ipv4Instance(sin));
+    break;
```
Comment (mattklein123) [inline source/common/network/address_impl.cc]: nit: line can't be reached.

File: source/common/network/address_impl.cc [hunk 4]
```diff
@@ -8,6 +8,37 @@
...
+  }
+  case AF_INET6: {
+    ASSERT(ss_len == sizeof(sockaddr_in6));
+    const struct sockaddr_in6* sin6 = reinterpret_cast<const struct sockaddr_in6*>(&ss);
+    ASSERT(AF_INET6 == sin6->sin6_family);
+    return InstanceConstSharedPtr(new Address::Ipv6Instance(*sin6));
+    break;
```
Comment (mattklein123) [inline source/common/network/address_impl.cc]: nit: line can't be reached.

File: source/common/network/address_impl.cc [hunk 5]
```diff
@@ -8,6 +8,37 @@
...
+  }
+  default:
+    throw EnvoyException(fmt::format("Unexpected family in getsockname result: {}", ss.ss_family));
+    return nullptr;
+  }
+  return nullptr;
```
Comment (mattklein123) [inline source/common/network/address_impl.cc]: nit: Would just do NOT_REACHED here if compiler complains

File: source/common/network/listen_socket_impl.cc [hunk 6]
```diff
@@ -15,6 +15,11 @@ void ListenSocketImpl::doBind() {
     throw EnvoyException(
         fmt::format("cannot bind '{}': {}", local_address_->asString(), strerror(errno)));
   }
+  if (local_address_->ip()->port() == 0) {
```
Comment (mattklein123) [inline source/common/network/listen_socket_impl.cc]: doBind() is also called in UdsListenSocket case so I think this line will crash without checking if ip() is not null or type is IP

Comment (htuch) [inline source/common/network/listen_socket_impl.cc]: Yeah, didn't run the full tests before pushing, will fix.

**Comment by htuch** [comment]
Fixed, tests pass (this is what happens when you write code while in the middle of a meting ;) )

File: source/common/network/listen_socket_impl.cc [hunk 7]
```diff
@@ -15,6 +15,12 @@ void ListenSocketImpl::doBind() {
     throw EnvoyException(
         fmt::format("cannot bind '{}': {}", local_address_->asString(), strerror(errno)));
   }
+  if (local_address_->type() == Address::Type::Ip && local_address_->ip() != nullptr &&
```
Comment (mattklein123) [inline source/common/network/listen_socket_impl.cc]: nit: local_address_->ip() != nullptr is guaranteed if local_address_->type() == Address::Type::Ip

**Pull request merged**